### DISCUSSION TITLE
destOverride到底是干什么用的!??

### DISCUSSION THREAD

**Discussion by k79e** [v2fly/v2ray-core#903] [General]
> destOverride: ["http" | "tls" | "fakedns"]

说明书的解释让人看不懂啊
那个**重置**是什么意思!??

不开这个 或者把他设定为空白 还是能封锁域名啊

有人能解答下么?
2021-04-14 17:40:18
更新一下 我知道那个fakedns要把ip转换回去域名 
我想问的是 除了fakedns之外那2个选项干啥的

2021-04-17 16:51:21
如开发者所说 就是把IP转回去域名 
fakedns以外具体干啥的我也想不到啊 看来就只有透明代理+路由了?

2021-04-17 21:40:55
看来也就fakedns和透明代理需要这个了 其他情况应该没必要

**Comment by Cwek** [comment]
sniffing的destOverride？
不就是流量目标地址探测？就是一些以IP作为目标请求的连接，通过对应的应用层探测处理，找回对应的目标域名并改成以相应的域名作为目标请求。主要是针对一些像将IP转回域名访问的情况？

**Reply by k79e** [reply]
走ip访问还要域名干啥!??

**Reply by k79e** [reply]
ip跳回域名和加https都一样的 是客户端收到跳转 然后重新发起连接 
这跟那个功能没什么关系吧.

**Reply by Cwek** [reply]
有些请求想通过域名来匹配路由（或者说解析出来的地址不理想，例如ss-Android，直接访问过来的请求都是IP地址），就需要将通过以IP地址作为目标的请求转回域名。

**Reply by Cwek** [reply]
如果现在有fakedns的话（fakedns替换目标地址的部分由已有的sniffing功能实现），泛用性更大，因为可以与应用层的域名嗅探脱钩，由dns控制访问目标地址，再将fakedns用目标地址劫持回真正的目标请求域名。

**Reply by k79e** [reply]
> 有些请求想通过域名来匹配路由（或者说解析出来的地址不理想，例如ss-Android，直接访问过来的请求都是IP地址），就需要将通过以IP地址作为目标的请求转回域名。

咋ss又冒出来了!??

**Reply by k79e** [reply]
> 如果现在有fakedns的话（fakedns替换目标地址的部分由已有的sniffing功能实现），泛用性更大，因为可以与应用层的域名嗅探脱钩，由dns控制访问目标地址，再将fakedns用目标地址劫持回真正的目标请求域名。

fakedns本身就有ip和域名之间的映射表 没必要开那个吧? 如果是那样 destoverride就是fakedns自己的特性 跟其他没关系了.

**Reply by Cwek** [reply]
> > 有些请求想通过域名来匹配路由（或者说解析出来的地址不理想，例如ss-Android，直接访问过来的请求都是IP地址），就需要将通过以IP地址作为目标的请求转回域名。
> 
> 咋ss又冒出来了!??

因为我手机客户端是用ss-ardroid，而服务端是v2的ss模块。

**Reply by Cwek** [reply]
> > 如果现在有fakedns的话（fakedns替换目标地址的部分由已有的sniffing功能实现），泛用性更大，因为可以与应用层的域名嗅探脱钩，由dns控制访问目标地址，再将fakedns用目标地址劫持回真正的目标请求域名。
> 
> fakedns本身就有ip和域名之间的映射表 没必要开那个吧? 如果是那样 destoverride就是fakedns自己的特性 跟其他没关系了.

以v2为中间，左为下游接入，右为上游请求。

fakedns和sniffing解决的是左边将IP转回域名的问题，destOverride是解决右边谁去处理请求资源时的域名解析问题（是使用回v2的dns配置还是让操作系统处理）

**Comment by Loyalsoldier** [comment]
简单说就是，将 V2Ray 内的 DNS 模块解析到的 IP 地址（可能是真实 IP 地址，也可能是 Fake DNS 分配的“假 IP” 地址），替换回原本请求的域名，再发往服务端。这也就是你之前吐槽说，本地客户端 DoH 已经获取到 IP 地址了，为啥还需要服务端发起一次 DNS 查询的原因。

因为服务端收到的流量的目标地址被重新替换回了域名。

**Reply by k79e** [reply]
我之前不是freedom没配置好么... 那现在freedom不管也可以了!??

**Reply by Cwek** [reply]
freedom的domainStrategy？
AsIs相当于直接通过域名打开socket，域名解释由操作系统（/etc/resolv.conf）解决
UseIP就是由v2配置的DNS解析域名后再通过IP地址打开socket。

**Reply by k79e** [reply]
所以这个只跟fakedns有关? 跟其他没任何关系!?? 那个不是很必要么 fakedns没这个岂不是就废了!??
但是他有3个选项 那fakedns之外的2个干啥的?

**Comment by k79e** [comment]
额 重新表明一下 我想知道是fakedns之外的应用 http/tls这2个设置干啥用的?
他能把这个ip转换为域名有啥用?

**Reply by xiaokangwang** [reply]
用于规则匹配，rule里面有域名规则。

**Reply by xiaokangwang** [reply]
目的都是把IP转换回域名，每个选项是不同的实现思路。

**Reply by xiaokangwang** [reply]
哪个适合自己选哪个

**Reply by k79e** [reply]
看你这解释我更疑惑了 那你这意思 可以用fakedns的时候 那个选项不填fakedns 填别的也可以!??

**Reply by k79e** [reply]
> 用于规则匹配，rule里面有域名规则。

我把那个设置打开 然后里面填空的 域名规则还有效哇

**Reply by lanchengjiu** [reply]
HTTP/S协议的首部内容比较多，包含了域名，所谓的嗅探，就是把这个域名给提取出来，这样在之后各种rules匹配中，就可以使用域名进行匹配了。这个嗅探技术只支持HTTP和HTTPS，这块就是说遇到集合中包含的协议，就执行嗅探，所以一般这两个项都会配置进去。

**Reply by k79e** [reply]
> 这样在之后各种rules匹配中，就可以使用域名进行匹配了

不开嗅探照样能用路由

**Reply by ygcaicn** [reply]
在op的透明代理下就很好理解了，局域网设备浏览器要访问google.com，首先向op请求dns得到一个ip（op无论用什么方式查询dns，比如简单的全部返回一个10.0.0.10，只要记录这个ip用于后续流量转发就行）然后浏览器发起ip socket连接，op的iptables将这个连接的流量发给V2ray，这时destOverride根据协议，比如http/s，嗅探数据包内容，将目标地址该写为域名再发送到V2ray的服务端。

**Reply by k79e** [reply]
> 在op的透明代理下就很好理解了

op是什么? 以前测试tproxy的时候也没开嗅探 正常能用哇

**Reply by ygcaicn** [reply]
> > 在op的透明代理下就很好理解了
> 
> op是什么? 以前测试tproxy的时候也没开嗅探 正常能用哇

OpenWrt。没开destOverride能正常用说明google.com得到了正确的dns响应，所以才正常。参考上面的极端情况把要代理的域名全部返回10.0.0.10，然后把目的地址为10.0.0.10的流量全部转发给V2ray，这就体现出destOverride的作用了。

**Reply by lanchengjiu** [reply]
我记得正常浏览网页是先发DNS解析请求，拿到结果时候在建立TCP连接。但是浏览器使用socks5将流量发给v2ray的时候，会把域名一起发过去，和正常流程不一样，v2ray就能直接拿到域名信息了，可能是这个原因吧。嗅探功能好像更多是用于路由器做透明代理的情景下

**Reply by ygcaicn** [reply]
把要代理的域名全部返回10.0.0.10，这就是PassWall中的“Fake IP”DNS模式的方案。
![image](https://user-images.githubusercontent.com/17684070/115016582-d13cf880-9ee7-11eb-99f3-300e8a51d8cc.png)

但是这样V2ray在嗅探时只能重写http/s的流量，其它例如http2的流量这样就无能为力了。

**Reply by ygcaicn** [reply]
> 我记得正常浏览网页是先发DNS解析请求，拿到结果时候在建立TCP连接。但是浏览器使用socks5将流量发给v2ray的时候，会把域名一起发过去，和正常流程不一样，v2ray就能直接拿到域名信息了，可能是这个原因吧。嗅探功能好像更多是用于路由器做透明代理的情景下

socks5的情况理论上完全无需destOverride，但是也不排除某些软件在发送socks5流量之前自行的解析，记得早期Chrome有这种行为。

**Comment by k79e** [comment]
> 但是这样V2ray在嗅探时只能重写http/s的流量，其它例如http2的流量这样就无能为力了。

嗅探不支持h2么!?? 不应该把 他包括在tls里面的
其他人有这个情况么?

**Reply by ygcaicn** [reply]
哈哈我的过，不严谨的表述。仅仅因为我在这样使用时github以及其它使用h2的网站无法打开。。

**Comment by k79e** [comment]
tor流量走过去 全变成域名访问了 然后tor废了..................
怎么会这样呢 是不是识别错误

**Reply by lanchengjiu** [reply]
嗅探功能和tor有冲突，这个应该算很常见的问题了，可以谷歌下相关内容很多

**Comment by k79e** [comment]
这是把公网dst ip里面包含的sni信息用来做连接了
fakedns / 透明代理 就没公网ip吧

再说那个是重置功能啊 他可以关闭 跟嗅探有什么关系
嗅探出域名然后用路由去 也没问题啊.
然后就变成另一个问题了 关了那个 嗅探好像失效了
https://github.com/v2fly/v2ray-core/discussions/914

**Comment by k79e** [comment]
说明书好像没提到
如果 destOverride: []
空白放那里 等于关闭嗅探吧.

这样配置之后 好像嗅探不出域名了.... 这到底是不是bug呢
